### PR TITLE
libxml2: 2.9.14 → 2.10.0

### PR DESCRIPTION
**Pull request by jtojnar**
https://gitlab.gnome.org/GNOME/libxml2/-/compare/v2.9.14...v2.10.0

- Massive cleanups
- CVE-2022-2309
- More security fixes

<del>

Unfortunately, one of the clean ups means using Automake for finding Python,
which does not support Python 3.10:

    checking for PYTHON... no
    configure: error: Package requirements (python-3.1) were not met:

    No package 'python-3.1' found

https://gitlab.gnome.org/GNOME/libxml2/-/issues/392

We have to patch it. Doing it locally to reduce the rebuild time since this is security critical.

</del>

Python detection was further broken by strictDeps for unknown reasons.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/libxml2/default.nix [hunk 1]
```diff
@@ -47,11 +50,22 @@ stdenv.mkDerivation rec {
...
-
   nativeBuildInputs = [
     pkg-config
-    autoreconfHook
+    (autoreconfHook.override {
+      automake = automake.overrideAttrs (attrs: {
+        patches = attrs.patches or [ ] ++ [
+          # Unable to find Python 3.10
+          (fetchpatch {
+            url = "https://github.com/autotools-mirror/automake/commit/930a9a73f4bb776ec334eff4cf6e182802841daa.patch";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libxml2/default.nix:60]: Please love this to automake. This is a general, annoying issue.

Comment (jtojnar) [reply pkgs/development/libraries/libxml2/default.nix]: I am unable to reproduce it any more.

**Comment by mweinelt** [comment]
Breaks postgresl_14' xml test. cc @marsam

File: pkgs/development/libraries/libxml2/default.nix [hunk 2]
```diff
@@ -47,8 +48,6 @@ stdenv.mkDerivation rec {
     ./utf8-xmlErrorFuncHandler.patch
   ];
 
-  strictDeps = true;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libxml2/default.nix:50]: Maybe this can now be added back? Also maybe it looks for the python in nativeBuildInputs?

Comment (jtojnar) [reply pkgs/development/libraries/libxml2/default.nix:50]: No, that still does not work.

**Comment by jtojnar** [comment]
Just the xml tests failing:

```
     xml                          ... FAILED      471 ms
```
[regression.out](https://github.com/NixOS/nixpkgs/files/9384727/regression.out.txt)
[regression.diffs](https://github.com/NixOS/nixpkgs/files/9384739/regression.diffs.txt)

**Comment by jtojnar** [comment]
Relevant source code:
- https://gitlab.gnome.org/GNOME/libxml2/-/blob/d54f829ff7aed095668547705f32bb9c7545b67e/parserInternals.c#L1131

- https://github.com/postgres/postgres/blob/3097bde7dd1d9e7b1127b24935e70ca4609e22b2/src/test/regress/sql/xml.sql#L68
- https://github.com/postgres/postgres/blob/3097bde7dd1d9e7b1127b24935e70ca4609e22b2/src/backend/utils/adt/xml.c#L756

Managed to create a repro:

```nix
with import <nixpkgs> { };
with pkgs;

let
  testSrc = writeText "test.c" ''
    #include <libxml/chvalid.h>
    #include <libxml/parser.h>
    #include <libxml/parserInternals.h>
    #include <libxml/tree.h>
    #include <libxml/uri.h>
    #include <libxml/xmlerror.h>
    #include <libxml/xmlversion.h>
    #include <libxml/xmlwriter.h>
    #include <libxml/xpath.h>
    #include <libxml/xpathInternals.h>


    int main(int argc, char const *argv[]) {
        xmlChar *string = "<undefinedentity>&idontexist;</undefinedentity>";
        xmlParserCtxtPtr ctxt = NULL;
        xmlDocPtr doc = NULL;

        xmlInitParser();

        ctxt = xmlNewParserCtxt();
        if (ctxt == NULL) {
            printf("could not allocate parser context\n");
            return 1;
        }

        doc = xmlCtxtReadDoc(ctxt, string, NULL, "UTF-8", XML_PARSE_NOENT);
        if (doc == NULL)
        {
            printf("invalid XML content\n");
            return 1;
        }

        return 0;
    }
  '';
in runCommand "libxml2-collate-test" {
    buildInputs = [ libxml2 pkg-config gcc ];
  } ''
    cc \
      $(pkg-config --cflags libxml-2.0) \
      $(pkg-config --libs libxml-2.0) \
      ${testSrc} -o test

    mkdir -p $out
    mv test $out/test

    env -i $out/test
  ''
```

**Comment by jtojnar** [comment]
Opened https://gitlab.gnome.org/GNOME/libxml2/-/issues/397, where I bisected it to https://github.com/GNOME/libxml2/commit/aab584dc3127edf383c3d1c6acda2b83353a7150.

**Comment by jtojnar** [comment]
PostgreSQL build was fixed by cherry-picking libxml2 patch.

**Comment by SuperSandro2000** [comment]
@ofborg build [postgresql](https://search.nixos.org/packages?channel=unstable&from=0&size=50&buckets=%7B%22package_attr_set%22%3A%5B%22No%20package%20set%22%5D%2C%22package_license_set%22%3A%5B%5D%2C%22package_maintainers_set%22%3A%5B%5D%2C%22package_platforms%22%3A%5B%5D%7D&sort=relevance&type=packages&query=postgres)

**Comment by risicle** [comment]
At least the CVE is only a null pointer dereference.

**Comment by risicle** [comment]
We applied the lxml workaround for stable in #182708, but it may be worth applying https://gitlab.gnome.org/GNOME/libxml2/-/commit/5930fe01963136ab92125feec0c6204d9c9225dc too in case any other apps are using a similar pattern.

**Comment by jtojnar** [comment]
That commit is already part of this update.

**Comment by risicle** [comment]
I'm discussing stable.

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**